WS-Discovery interface implementation

fuzzard opened the pull request:
## Description
This implements the WS-Discovery protocol for SMB discovery for posix platforms

This PR is predominantly to get early feedback, and to have others test and let me know of any obscure cases we might need to handle for a more robust implementation. We essentially stick to the bare minimum that i believe is required to adhere to the WSD specs and act as a client.

Be gentle on the review ;) I have no doubts you guys will have some ideas about better implementations. Let me know, open to anything suggested.

ToDo:

- [X] Interface for WSD to unify Windows/posix creation/init for their respective WSD implementations
- [X] Settings to allow user to enable/disable the WSD daemons if they wish
- [ ] Further testing
- [x] More code comments
- [ ] ~~ipv6 support - Most likely a followup PR~~ - Do in another PR some day

## Motivation and context
Provide detection of SMB servers as per WS-Discovery specification

## How has this been tested?
OSX client, Synology Nas + windows smb shares
Android aarch64 client tested working on same network

## What is the effect on users?
Allows users the ability to have their system detect smb servers that adhere to WS-Discovery spec

Fixes #18840 

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[posix] WS-Discovery Implementation

Implements WS-Discovery interface
Implements Posix WS-Discovery daemon
Add WS-Discovery interface settings and start/stop to NetworkServices

add component logging for WS-Discovery

garbear commented on xbmc/ServiceManager.cpp:
```diff
@@ -37,6 +40,8 @@
 #include "utils/log.h"
 #include "weather/WeatherManager.h"
 
+#include <memory>
```
Not needed, this is included in the .h file

garbear commented on xbmc/platform/posix/filesystem/SMBWSDiscovery.cpp:
```diff
@@ -0,0 +1,96 @@
+CWSDiscovery::CWSDiscovery()
+{
+  // Set our wsd_instance ID to seconds since epoch
+  auto epochduration = system_clock::now().time_since_epoch();
```
are you sure you want system_clock and not stable_clock?

fuzzard commented:
Have advanced a bit further with this.

Settings to enable/disable is now in place

Have had a crack at implementing an interface for WS-Discovery. The posix stuff in here uses it, i'll look to see how quickly i can get the existing windows one to use it. Should just be a couple class/method renames, so dont think it'll be too much trouble.

This now use network services to start/stop and handle the settings handler for start/stop. No longer a need for platform specific initstagethree.

Would love any feedback from posix platform testers (android, linux, apple) about whether it works in your network environment.

Main "feature" remaining is the support for ipv6 networks. This currently only works with ipv4. I dont think i'll look to get ipv6 working in this PR, dont want to have it drag along too much

Commit:
[Windows][WS-Discovery] migrate windows to IWSDicovery interface and WSDiscovery namespace

Migrate existing Windows WS-Discovery service to the new WS-Discovery Interface

Make Settings SMB client section visible for Windows but hide the Samba only settings on Windows that are used on posix platforms only.

phunkyfish commented:
I'll test once the build is finished. Also have mac + synology. Thanks for doing this, WS discovery was much needed for the non windows plaforms.

phunkyfish commented:
I didn't have any luck with this. I went to browse Windows Network (SMB), got one address 192.168.1.64 clicked it and got software caused connection abort. My synology did not appear in the list.

fuzzard commented:
Ill ask the obvious first question, got WS-Discovery enabled on the synology?

phunkyfish commented:
DS version needed to be updated. Work perfectly after that. I wouldn't let IPV6 hold this up. I don't think it's common to run it home networks.

fuzzard commented:
Jenkins build this please

fuzzard commented:
Just pinging @thexai . If you have time would you give it a runtime test on windows. Seemed to work locally, and the windows changes are very minimal, so not expecting too much trouble.

thexai commented:
Runtime tested on Windows and WDS works fine as before. Very good job unifying posix and Windows interfaces.

enen92 commented on xbmc/network/NetworkServices.cpp:
```diff
@@ -624,6 +648,11 @@ bool CNetworkServices::StartServer(enum ESERVERS server, bool start)
       break;
 
+    case ES_WSDISCOVERY:
+      // the callback will take care of starting/stopping zeroconf
```
Typo from copy/paste, you likely mean wsdiscovery on the comment and not zeroconf

fuzzard commented:
> Runtime tested on Windows and WSD works fine as before. Very good job unifying posix and Windows interfaces.
> 
> NOTE: There are no GUI settings to disable WS-Discovery in Windows although this is the same as before.

So the setting should be there for windows. Where I've put it is inside Services -> Smbclient and the new group inside there for WS-Discovery. The SMBclient settings is available at lvl2, I think that's advanced.

I'm open to moving this, however I think default to on, and as SMB is very common, for most users it'll be ok. I've just added the option to allow more advanced users to disable it if they wish.

thexai commented:
Default on is OK for me too. 

> So the setting should be there for windows

What I am say is that setting is not visible on Windows because entire SMB settings are not visible:

https://kodi.wiki/view/Settings/Services
SMB Client (not visible for Windows users)

It might make sense to put WS-Discovery separate since it is not unique to SMB. Other services or protocols could make use of it (WS-Discovery only provides a list of IP's).

jjd-uk commented:
@thexai is all that is missing for Windows the presentation of the settings in the GUI? If so I can take a look at the setting xml's as I've done work on them before, should just require some dependencies in the settings.xml to be changed.

thexai commented:
@jjd-uk, it's not necessary: I think that hiding the settings in Windows is done on purpose since those settings are for Samba. In Windows, Samba is not used, but a native SMB implementation. Those settings would still have no effect on Windows.

jjd-uk commented:
That shouldn't be an issue, we should be able to make the SMB client section available to all platforms so the ws-discovery settings can be seen, then just hide the samba related settings that do nothing on Windows.

jjd-uk commented:
To show the setting on Windows (but not the samba related stuff) then this is all that is needed https://github.com/jjd-uk/xbmc/commit/60c97e08d72f23139eb1f12660583f4ab5bd7fa0

So on Windows this is displayed

![image](https://user-images.githubusercontent.com/5781142/126188384-d325ba3d-1af8-4198-99dc-fe1c8bfd42ad.png)

Haven't tested whether that setting does anything on Windows yet however.

jjd-uk commented:
@fuzzard apologies, I seems to have pushed the commit to both the branch on my repo and also yours. Oops didn't mean for that to happen, hopefully not messed anything up.

thexai commented on xbmc/platform/win32/filesystem/Win32SMBDirectory.cpp:
```diff
@@ -637,14 +638,16 @@ static bool localGetShares(const std::wstring& serverNameToScan, const std::stri
 static bool localGetServers(const std::string& urlPrefixForItems, CFileItemList& items)
 {
-  auto wsd = CWSDiscoverySupport::Get();
+
```
Very minor: this extra blank line is not needed

fuzzard commented:
> @fuzzard apologies, I seemed to have pushed the commit to both the branch on my repo and also yours. Oops didn't mean for that to happen, was only meant to go to my repo, hopefully I've not messed anything up.

No problems at all, that's perfect. Was what I would have gone looking for. 

@thexai has a point about WS-Discovery not being SMB specific, however for our purposes it is intrinsically tied to it, so still I think the best place for that setting for now.

jjd-uk commented:
Yep that was my thought too. If further services are added that use WS-Discovery, then we could consider moving the setting at that point to General or wherever.

fuzzard commented:
Alright, i consider this ready now. Have added a bit more failure checking/cleanup for posix. Added comments and assorted fixups.
Happy with it now if people want to do a last final review.

thexai commented:
Tested with the settings modification and I can confirm that is shown now and works as expected: enable/disable WS-Discovery works fine on Windows.

reardonia commented:
On posix, this is redundant to the service provided by the name resolver in systemd, ie systemd-resolve, right?  LLMNR is preferred to WS-D, and is the default resolver on Windows 10, I thought.

fuzzard commented:
It's just another protocol for service discovery. Setting is there to disable if you don't need it

jernejsk commented:
Last time I read about LLMNR, it was discouraged due to security issues (https://www.blackhillsinfosec.com/how-to-disable-llmnr-why-you-want-to/).

fuzzard commented:
My understanding of llmnr is it's not there for service detection, this is more equivalent to dns-sd. Llmnr works finding clients in non DNS situations, but doesn't know anything about what the client offers as a service.
So in this instance, sure you get a list of every network device, but how would you know what offers some sort of service Kodi cares about?

reardonia commented:
> Last time I read about LLMNR, it was discouraged due to security issues (https://www.blackhillsinfosec.com/how-to-disable-llmnr-why-you-want-to/).

No, that link is just wrong. It's not discouraged for security reasons.

In fact, systemd enables it by default, on Linux. 

LLMNR and mDNS are both fine; I expect LLMNR to be deprecated mostly because Microsoft decided to support mDNS, but not for security reasons.  W-DS has as many or more security issues.

reardonia commented:
> It's just another protocol for service discovery. Setting is there to disable if you don't need it

Just to be clear: this is just for _browsing_ a network?  Not for the lookup of a known server name? I don't think the service-type advertisement actually matters, since every Windows machine is technically also a server, so you don't get much benefit there.

fuzzard commented:
Im really not sure what your after?

This implements smb service discovery using WS-Discovery protocol. It finds devices that offer that discovery method on a network. Nothing more, nothing less. If you dont use WS-Discovery, turn it off. If you do, turn it on.

By all means, this isnt a complete implementation of WS-D, as it doesnt need to be for our purposes inside kodi.

Posix also isnt linux only, and definitely isnt systemd only. Android, Apple, linux, freebsd platforms all use the posix defines/implementations in kodi as a base for their platforms.

mglae commented on xbmc/platform/posix/filesystem/SMBWSDiscoveryListener.cpp:
```diff
@@ -0,0 +1,510 @@
+    "http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches";
+static const std::string WSD_ACT_PROBE = "http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe";
+static const std::string WSD_ACT_RESOLVE =
+    "https://schemas.xmlsoap.org/ws/2005/04/discovery/Resolve";
+static const std::string WSD_ACT_RESOLVEMATCHES =
+    "https://schemas.xmlsoap.org/ws/2005/04/discovery/ResolveMatches";
```
```suggestion
    "http://schemas.xmlsoap.org/ws/2005/04/discovery/Resolve";
static const std::string WSD_ACT_RESOLVEMATCHES =
    "http://schemas.xmlsoap.org/ws/2005/04/discovery/ResolveMatches";
```
Actions start with "http://". [wsdd2](https://github.com/Andy2244/wsdd2) is complaining about unsupported queries.

fuzzard replied:
Interesting that win10 and Synology both return https. I'll add a check for either, thanks for the testing

fuzzard replied:
had a chance to look at it tonight, and now i see that it was just the resolve/matches actions i had C&Ped the urls in wrong. Should be fixed now.

mglae commented on xbmc/platform/posix/filesystem/SMBWSDiscovery.cpp:
```diff
@@ -0,0 +1,110 @@
+    // delim1 used to strip protocol from xaddrs
+    // delim2 used to strip anything past the port
+    const std::string delim1 = "://";
+    const std::string delim2 = ":5357";
```
Unicast SOAP HTTP WS-Dicsovery responder are allowed to listen on any TCP port, e.g. wsdd2 is using 3702.

fuzzard replied:
Can you show me a log for an instance of this?
This was very much specific to what I could test (win10, Synology wsd), so if you can give me a log so I can see how I need to handle the xaddrs value, that'd be great.

mglae replied:
Log when using the "https:" actions: http://ix.io/3ulQ. Neither the router nor the wsdd2 device is listed.

With "http:" actions both devices are detected: http://ix.io/3ulS. Wsdd2 is using 192.168.11.3:3702

fuzzard replied:
Cheers, i'll look to generalise the xaddrs stripping a bit more hopefully this weekend.
I'll work out a way to handle the actions as well. According to the spec the "actions" can effectively be disregarded, which might be why windows and wsd.py implementations might not care about what is sent for that part of the SOAP xml packet, and hence they work with this implementation.

fuzzard replied:
Have altered the delimiter matching to match on the colon instead of the :5357. Id expect it to be ok as that, but give it a test and let me know how it goes.

As part of this ive also moved the WS-Discovery logging to a new component logging. I figure for the most part we dont really care about WS-D unless we want to debug it, so a user can turn on component logging to get the WS-D debug stuff.
With that, ive added some extra logging in states to provide some info about the received packets that are discarded for various reasons. Should give more info in case people come across other setups in the future.

mglae replied:
Looks good. All devices are visible now.

fuzzard replied:
Thanks for testing, greatly appreciated.

mglae commented:
To open a wish list I do prefer to see the host names instead of IP addresses. No need to add the feature to this PR.

In addition Windows 10 is requiring credentials for SMB shares. Choosing the IP address in the selector result in "Parameter error" and the machine cannot be selected. Again nothing to be fixed in this PR.

fuzzard commented:
This just shows whatever is provided by the WS-D search. If it returns the xaddr as a hostname, it'll show as a hostname (Synology does this for me at least)

No idea about the windows creds side of things. Maybe need to add to passwords.xml or something like that.

jjd-uk commented:
Windows credentials should no longer to be required following https://github.com/xbmc/xbmc/pull/19599

fuzzard commented:
I think mglae's usecase is linux to windows smb server, so that wouldnt have an effect on that situation

Commit:
[WSDiscovery][Windows] direct WSD logging to WSDicovery component logging

fuzzard commented:
one final ping for @thexai Ive added a commit to move the windows implementation to use the "Component Logging" aspect for WS-Discovery logging.
If you dont want the windows stuff to be tied to the WS-Discovery component logging, let me know and i can drop it.

My reasoning is if someone cares about the logging details for WS-Discovery, they can turn it on, otherwise wont be (other than start/stop daemon) logged in a normal debug log. Less un-targeted logging, and allows implementation to provide more verbose logging if it benefits us

thexai commented:
Seem right. Perhaps only for Windows it was not necessary due to the small amount of logs but is good that it is the same for all platforms.

thexai commented on xbmc/platform/win32/network/WSDiscoveryWin32.cpp:
```diff
@@ -138,7 +139,7 @@ HRESULT STDMETHODCALLTYPE CClientNotificationSink::SearchFailed(HRESULT hr, LPCW
 
   // This must not happen. At least localhost (127.0.0.1) has to be found
-  CLog::Log(LOGWARNING,
+  CLog::Log(LOGWARNING, LOGWSDISCOVERY,
```
I would like to leave this log visible as it was before since it is something that normally never has to happen, so it is a warning that something is wrong.

fuzzard replied:
no worries, will change it back

StormTrooper commented:
Look like there is a typo that crept in with the change on NetworkServices.cpp

```CSettings::SETTING_SERVICES_WSDISCOVERY,```

The comma at the end needs to be dropped

fuzzard commented:
No typo, it's intended. For an explanation see https://stackoverflow.com/questions/7043372/int-a-1-2-why-is-a-trailing-comma-in-an-initializer-list-allowed

mglae commented:
Debug crash log at http://ix.io/3BTW

fuzzard commented:
Think ive found a way to replicate this fault on ios. Seems to be a race condition. Networkservices message is received before the servicebroker instance is started in initstagethree().

My quick and dirty was to relocate https://github.com/xbmc/xbmc/blob/973c5a584c3696a5e61e399325540cd7abecdc56/xbmc/ServiceManager.cpp#L195-L197 to somewhere in initstagetwo.

If i relocate it even to the start of initstagethree i was able to continue on with ios boot, but on a different system maybe even that might fail due to timing.
I'll have a chat to some others and see if i can work out whats best.
